islutils/access_patterns.cc

```diff
@@ -58,23 +58,32 @@ isl::map SingleInputDim::transformMap(isl::map map,
 ///////////////////////////////
 
 static isl::map mapToNext(isl::space space) {
-  using map_maker::operator==;
-  int dim = space.dim(isl::dim::set);
 
-  auto result = isl::map::universe(space.map_from_set());
-  if (dim == 0) {
-    return result;
-  }
+  space = space.map_from_set();
+  isl::map map = isl::map::universe(space);
+  unsigned lastDimension = map.dim(isl::dim::in) - 1;
+
+  // Set all but the last dimension to be equal for the input and output
+  //
+  //   input[i0, i1, ..., iX] -> output[o0, o1, ..., oX]
+  //     : i0 = o0, i1 = o1, ..., i(X-1) = o(X-1)
+  for (unsigned i = 0; i < lastDimension; ++i)
+    map = map.equate(isl::dim::in, i, isl::dim::out, i);
+
+  // Set the last dimension of the input to be strict smaller than the
+  // last dimension of the output.
+  //
+  //   input[?,?,?,...,iX] -> output[?,?,?,...,oX] : iX < oX
+  map = map.order_lt(isl::dim::in, lastDimension, isl::dim::out, lastDimension);
```

Comment:
Consider the ```StrideTest``` in the test_access.cc. If I use your function ```mapToNext``` I got a delta that looks like ``` delta{[i0] : false}``` while in the proposed PR the delta looks like ``` [i0] : (i0) mod 2 and 0 < i0 < 1022}```. So basically I was not able to get the delta I wanted.

Reply:
I'm more interested in how this solution is correct for all cases (and also understanding why mine was not), rather than in how it fixed one particular case to get the result you wanted (and also why the result you wanted is correct).

My, apparently wrong, reasoning was mostly driven by the idea of consecutivity: take two consecutive points in the schedule space (add 1 to the last dimension of the output space) and look which points in the address space they access.  If the points are also consecutive, then we have stride 1, if they are all two elements apart, we have stride 2, and so on.  Now, it does _not_ work if there are points in schedule space that access nothing at all, which is the case when you have a non-dense domain with identity schedule.  In your example, there is loop step 2, so any odd iteration is accessing an empty set of elements, so the relation between array elements between two consecutive points in schedule space is also empty (at least one of the two points corresponds to an odd iteration, accessing an empty space of elements).  I'd like you to explain me how relating a point in iteration space to _all_ its successors and the other changes correspond to the accessed elements.

Reply:
I can also argue that, in your example, the stride is indeed not 2.  It depends on what we understand by access stride: is it a coefficient in the subscript as in `A[2*i]`, is it the distance in accessed elements independently of how it is expressed in the code?

Reply:
Why in the example the stride is not 2? For me the stride is the distance between two accessed elements. For example, ```A[2*i]``` with i incremented by 1 has stride 2. In the example both A and B have stride 2.

Reply:
The key thing is your reply is "for me" :)  There are two possibilities, and there are practical cases where both would be applicable.  Unless we clearly define which one we chose, we won't agree.

I still need that high-level reasoning as to why, under the distance-between-accessed-elements definition, your approach computes something meaningful.

Reply:
Here is an example where you have a distance 2, but I'm not sure whether it should be interpreted as stride 2: `[t]->A[a]: a = (6*t) mod 8`.  If you are interested in the fact that, in general, every second element gets accessed, it's okay.  If, however, you need some information about _when_ these elements are accessed, this becomes trickier.

Reply:
>  Is this answering your question?

No, it does not, and your explanation does not seem to be correct.

> I mean, the function mapToNext creates a map such that the last dimension of the input is less then the last dimension of the output.

I can read the code :)  What I am missing is how this is interpreted in terms of schedules/accesses and so on.

>  you get the distance between two points in the schedule space.

Two problems here.

1. the map you construct with `mapToNext` connects a point in the schedule space on the left to _many_ points on the right, not just _two_.  Given `{[i]->[j]: i < j}`, for fixed `i=0`, all of `j=1,2,3,4,5,6,7,...` are in that map.  I'm struggling to understand why do you need all points.
2. I don't get the distance until a call to `deltas`.

Look at my reasoning above and try to explain (and understand yourself) what your code is doing using the same connection to higher-level concepts.  To start, you create a map between a point in the schedule space and all points that follow it in the last dimension with all previous fixed.  Then you transform this map into a map between sets of array elements accessed at the given points in time (you do get a _set_ of points rather than a single point on the right hand side because you had multiple points on the right hand side mapped from one point on the left hand side).  In general, I would expect the deltas to be `ax > 0` in most cases.